probe/endpoint/ebpf.go

```diff
@@ -0,0 +1,256 @@
+package endpoint
+
+import (
+	"bytes"
+	"encoding/binary"
+	"net"
+	"strconv"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	bpflib "github.com/iovisor/gobpf/elf"
+	"github.com/kinvolk/tcptracer-bpf/pkg/byteorder"
+	"github.com/kinvolk/tcptracer-bpf/pkg/offsetguess"
+)
+
+type eventType uint32
+
+// These constants should be in sync with the equivalent definitions in the ebpf program.
+const (
+	_ eventType = iota
+	EventConnect
+	EventAccept
+	EventClose
+)
+
+func (e eventType) String() string {
+	switch e {
+	case EventConnect:
+		return "connect"
+	case EventAccept:
+		return "accept"
+	case EventClose:
+		return "close"
+	default:
+		return "unknown"
+	}
+}
+
+// tcpEvent should be in sync with the struct in the ebpf maps.
+type tcpEvent struct {
+	// Timestamp must be the first field, the sorting depends on it
```

Comment:
tl;dr: it is implicitly required by the gobpf API we are using. We should update the API to make it more explicit... The cast is done here:
https://github.com/iovisor/gobpf/blob/a81bc05bf71ec17bf4c8d388621d2c81bc1f6a6b/elf/perf.go#L195

Longer version:

The kernel implements perf maps without locking to avoid contention between multiple execution of the ebpf program in parallel on multiple CPU. We actually have one perf file descriptor per CPU (each created with the syscall perf_event_open()).

But since each perf fd is independent from each other, we don't have any ordering guarantee between them. Other ebpf tracing programs in bcc currently don't mind that much about the lack of ordering. In practice, events are "very often" ordered.

Scope needs to process the events in chronological order, otherwise connections might appear to stay alive forever (e.g. if in a short-lived connection, a "close" event on one cpu is processed before its corresponding "connect" event on another cpu, the "close" event will be ignored because the connection is not know by Scope yet; then there will be the "connect" event without any following "close").

Perf events are opaque buffers. The perf subsystem on Linux does not handle reordering itself. The perf event struct is just a convention between the ebpf program ([struct tcp_ipv4_event_t](https://github.com/kinvolk/tcptracer-bpf/blob/b021c81f340f16376990ab9566ed1cace8ee9311/tcptracer-bpf.c#L25)) and Scope ([type tcpEvent struct](https://github.com/weaveworks/scope/pull/2135/files#diff-5023f232702c5a1b8177be985db66b08R40)).

However, we implemented the reordering in iovisor/gobpf, which is foreign to that convention. iovisor/gobpf does not know about the struct tcp_ipv4_event_t. So in the current implementation, it just assumes the ebpf program writes a timestamp at the beginning of the event buffer.